tokio/src/sync/lazy.rs

```diff
@@ -8,13 +8,18 @@ use std::panic::{RefUnwindSafe, UnwindSafe};
 use std::pin::Pin;
 use std::sync::atomic::{AtomicBool, Ordering};
 
+enum LazyUninitData<Fut, F> {
+    Function(ManuallyDrop<F>),
+    Future(ManuallyDrop<Fut>),
+}
+
 /// The inner data of a `Lazy`.
 /// Is `function` when not yet initialized,
 /// `value` with `Some` when initialized,
 /// and `value` with `None` when poisoned.
-union LazyData<T, F> {
-    value: ManuallyDrop<Option<T>>,
-    function: ManuallyDrop<F>,
+union LazyData<T, F, Fut> {
+    init: ManuallyDrop<Option<T>>,
+    uninit: ManuallyDrop<LazyUninitData<F, Fut>>,
```

Comment:
Perhaps we can move the semaphore field into the uninit variant?

```suggestion
    uninit: ManuallyDrop<Mutex<LazyUninitData<F, Fut>>>,
```

Reply:
Someone might still be awaiting on the semaphore (line 303) at the moment the cell is initialized. So I don't think that would work

Reply:
Thanks for the explanation!
I missed that particular case, though it would be great if we can remove that field since it is quite large and is unused once the cell is initialized